File: back-end/Domain/profit.go
package domain

// ProfitSummaryResponse holds aggregated profit metrics for a period
type ProfitSummaryResponse struct {
	TotalSales        float64 `json:"total_sales"`
	TotalExpenses     float64 `json:"total_expenses"`
	NetProfit         float64 `json:"net_profit"`
	Period            string  `json:"period,omitempty"`
}

// ProfitTrendDataPoint represents a single data point in the trends chart
type ProfitTrendDataPoint struct {
	Date          string  `json:"date"`
	TotalSales    float64 `json:"total_sales"`
	TotalExpenses float64 `json:"total_expenses"`
	NetProfit     float64 `json:"net_profit"`
}

// ProfitTrendsResponse holds an array of trend data points
type ProfitTrendsResponse struct {
	Trends []ProfitTrendDataPoint `json:"trends"`
	Period string                 `json:"period,omitempty"`
}

// ProfitCompareResponse compares two periods
type ProfitCompareResponse struct {
	Current   ProfitSummaryResponse `json:"current"`
	Previous  ProfitSummaryResponse `json:"previous"`
	ChangePct float64               `json:"change_pct"` // Percentage change in NetProfit
}

// ProfitQuery holds parameters for profit endpoints
type ProfitQuery struct {
	StartDate string `form:"start_date"`
	EndDate   string `form:"end_date"`
	Period    string `form:"period,default=daily"` // daily, weekly, monthly
}

File: web/lib/profit.ts
import { requestWithAuth } from "@/lib/api";

export type ProfitPeriod = "daily" | "weekly" | "monthly";

export type ProfitQueryParams = {
  businessId: string;
  startDate?: string;
  endDate?: string;
  period?: ProfitPeriod;
};

export type ProfitSummaryResponse = {
  total_sales: number;
  total_expenses: number;
  net_profit: number;
  period?: string;
};

export type ProfitTrendDataPoint = {
  date: string;
  total_sales: number;
  total_expenses: number;
  net_profit: number;
};

export type ProfitTrendsResponse = {
  trends: ProfitTrendDataPoint[];
  period?: string;
};

export type ProfitCompareResponse = {
  current: ProfitSummaryResponse;
  previous: ProfitSummaryResponse;
  change_pct: number;
};

const buildProfitQuery = (params: ProfitQueryParams) => {
  const query = new URLSearchParams({ business_id: params.businessId });

  if (params.startDate) {
    query.set("start_date", params.startDate);
  }

  if (params.endDate) {
    query.set("end_date", params.endDate);
  }

  if (params.period) {
    query.set("period", params.period);
  }

  return query;
};

export const fetchProfitSummary = (params: ProfitQueryParams) => {
  const query = buildProfitQuery(params);

  return requestWithAuth<ProfitSummaryResponse>(`/profit/summary?${query.toString()}`, {
    method: "GET",
  });
};

export const fetchProfitTrends = (params: ProfitQueryParams) => {
  const query = buildProfitQuery(params);

  return requestWithAuth<ProfitTrendsResponse>(`/profit/trends?${query.toString()}`, {
    method: "GET",
  });
};

export const fetchProfitComparison = (params: ProfitQueryParams) => {
  const query = buildProfitQuery(params);

  return requestWithAuth<ProfitCompareResponse>(`/profit/compare?${query.toString()}`, {
    method: "GET",
  });
};
